telemetry/src/lib.rs

```diff
@@ -0,0 +1,235 @@
+        cores: u32,
+    ) -> Self {
+        let store = TelemetryStore::new(instance_id, os, influx_version, storage_type, cores);
+        // TODO: decide the buffer size
+        let (sender, receiver) = mpsc::channel(100);
+        let telem_msg_handler = TelemetryMesageHandler::new(receiver, store.clone());
+        telem_msg_handler.handle_messages().await;
+        send_telemetry_in_background(store.clone()).await;
+        TelemetryHandle { sender }
+    }
+
+    pub async fn send(&self, msg: InternalTelemetryMessage) {
+        // TODO: handle failures in send
+        let _ = self.sender.send(msg).await;
+    }
+}
+
+/// The message types that can be sent through the `TelemetryHandle`.
+/// Currently we only need to collect numerical data periodically
+/// so just sticking with u64 as associated data is good enough
+pub enum InternalTelemetryMessage {
+    Writes(u64),
+    Reads(u64),
+    CpuUsage(u64),
+    CpuUtilization(u64),
+}
+
+/// This is responsible for receiving messages from `TelemetryHandle`
+struct TelemetryMesageHandler {
+    receiver: mpsc::Receiver<InternalTelemetryMessage>,
+    store: TelemetryStore,
+}
+
+impl TelemetryMesageHandler {
+    pub fn new(receiver: mpsc::Receiver<InternalTelemetryMessage>, store: TelemetryStore) -> Self {
+        TelemetryMesageHandler { receiver, store }
+    }
+
+    pub async fn handle_messages(mut self) {
+        tokio::spawn(async move {
+            loop {
+                if let Some(msg) = self.receiver.recv().await {
+                    // do something with msg
+                    self.store.handle_new_message(msg);
+                }
+            }
+        });
+    }
+}
+
+/// All of the internal messages are collected in this store
+#[derive(Clone)]
+struct TelemetryStore {
+    inner: Arc<sync::Mutex<TelemetryStoreInner>>,
+}
+
+impl TelemetryStore {
+    pub fn new(
+        instance_id: String,
+        os: String,
+        influx_version: String,
+        storage_type: String,
+        cores: u32,
+    ) -> Self {
+        let inner = TelemetryStoreInner::new(instance_id, os, influx_version, storage_type, cores);
+        TelemetryStore {
+            inner: Arc::new(sync::Mutex::new(inner)),
+        }
+    }
+
+    pub fn handle_new_message(&mut self, msg: InternalTelemetryMessage) {
+        match self.inner.lock() {
+            Ok(mut inner_store) => inner_store.handle_new_message(msg),
+            Err(e) => {
+                error!(error = ?e, "Cannot get the lock to add internal telemetry message");
+            }
+        };
+    }
+
+    pub fn snapshot(&self) -> Option<ExternalTelemetry> {
+        match self.inner.lock() {
+            Ok(inner_store) => Some(inner_store.snapshot()),
+            Err(e) => {
+                error!(error =?e, "Cannot get the lock to read from telemetry store");
+                None
+            }
+        }
+    }
+}
+
+/// This keeps copy of the all the samples sent as `InternalTelemetryMessage`
```

Comment:
Since we keep per minute rates, I would pre aggregate them into minute intervals. Then keep those summaries for the entire reporting period and get out whichever summaries are right then.

Reply:
Yea - that' s what I was thinking as well, I'll move it to do the aggregates/summaries in the next [issue](https://github.com/influxdata/influxdb/issues/25371)